### PR TITLE
BUG: IntegerArray.astype(boolean)

### LINKED ISSUE #31102: BUG: Cannot astype from IntegerArray to BooleanArray with missing values
For the boolean -> integer array conversion, I added a special path for that. But `astype`ing integer to boolean is currently not working:

```
In [23]: a = pd.array([1, 0, pd.NA])  

In [24]: a  
Out[24]: 
<IntegerArray>
[1, 0, <NA>]
Length: 3, dtype: Int64

In [25]: a.astype("boolean")  
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-25-41973ed53ee3> in <module>
----> 1 a.astype("boolean")

~/scipy/pandas/pandas/core/arrays/integer.py in astype(self, dtype, copy)
    454             kwargs = {}
    455 
--> 456         data = self.to_numpy(dtype=dtype, **kwargs)
    457         return astype_nansafe(data, dtype, copy=False)
    458 

~/scipy/pandas/pandas/core/arrays/masked.py in to_numpy(self, dtype, copy, na_value)
    124             ):
    125                 raise ValueError(
--> 126                     f"cannot convert to '{dtype}'-dtype NumPy array "
    127                     "with missing values. Specify an appropriate 'na_value' "
    128                     "for this dtype."

ValueError: cannot convert to 'boolean'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.

In [26]: a.astype(pd.BooleanDtype()) 
...
ValueError: cannot convert to 'boolean'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```

while for conversions to other nullable dtypes, this should be possible.

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/31102

### CODE REVIEW

File: pandas/tests/arrays/test_integer.py [hunk 1]
```diff
@@ -680,6 +680,13 @@ def test_astype_str(self):
         tm.assert_numpy_array_equal(a.astype(str), expected)
         tm.assert_numpy_array_equal(a.astype("str"), expected)
 
+    def test_astype_boolean(self):
+        # https://github.com/pandas-dev/pandas/issues/31102
+        a = pd.array([1, 0, None], dtype="Int64")
```
Comment (TomAugspurger) [inline pandas/tests/arrays/test_integer.py]: I've deliberately tested only 0 and 1 here. We should decide on a behavior for other values. Are all non-zero, non-NA values allowed and considered True?

Right now, BooleanArray._from_sequence doesn't allow other values

```python
In [2]: pd.array([1, 2], dtype='boolean')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-2a7683b5ab7f> in <module>
----> 1 pd.array([1, 2], dtype='boolean')

~/sandbox/pandas/pandas/core/construction.py in array(data, dtype, copy)
    282     if is_extension_array_dtype(dtype):
    283         cls = cast(ExtensionDtype, dtype).construct_array_type()
--> 284         return cls._from_sequence(data, dtype=dtype, copy=copy)
    285
    286     if dtype is None:

~/sandbox/pandas/pandas/core/arrays/boolean.py in _from_sequence(cls, scalars, dtype, copy)
    284         if dtype:
    285             assert dtype == "boolean"
--> 286         values, mask = coerce_to_array(scalars, copy=copy)
    287         return BooleanArray(values, mask)
    288

~/sandbox/pandas/pandas/core/arrays/boolean.py in coerce_to_array(values, mask, copy)
    169                 == values_object[~mask_values].astype(float)
    170             ):
--> 171                 raise TypeError("Need to pass bool-like values")
    172
    173     if mask is None and mask_values is None:

TypeError: Need to pass bool-like values
```

Comment (WillAyd) [reply pandas/tests/arrays/test_integer.py]: I would think want to follow Python semantics (?) but probably solved separate from this

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/tests/arrays/test_integer.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/test_integer.py]: I think there should be *some* way to go from "potentially not only 0's and 1's" to boolean, so I am fine here to follow Python and Numpy behavior for conversion (`bool(..)`, numpy's `astype`).

The question then is: should we follow this in the constructor, or are we fine with being more strict there? (I am personally fine with being more strict)

Comment (TomAugspurger) [reply pandas/tests/arrays/test_integer.py]: OK, added a test for non 0 / 1 values.

**Comment by TomAugspurger** [comment]
Added a test that IntegerArray.astype(boolean) works with non 0/1 values.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!

**Comment by jorisvandenbossche** [comment]
@MeeseeksDev backport to 1.0.x

**Pull request merged**